webrtc/browser-to-browser.md

```diff
@@ -0,0 +1,122 @@
+   _B_ creates an SDP offer via `RTCPeerConnection.createOffer()`.
+   _B_ initiates the signaling protocol to _A_ via the relayed connection from (1), see [Signaling Protocol](#signaling-protocol) and sends the offer to _A_.
+
+3. _A_ (outbound side of relayed connection) creates an `RTCPeerConnection`.
+   Again see [STUN](#stun) section on what STUN servers to configure at creation time.
+   _A_ receives _B_'s offer sent in (2) via the signaling protocol stream and provides the offer to its `RTCPeerConnection` via `RTCPeerConnection.setRemoteDescription`.
+   _A_ then creates an answer via `RTCPeerConnection.createAnswer` and sends it to _B_ via the existing signaling protocol stream (see [Signaling Protocol](#signaling-protocol)).
+
+4. _B_ receives _A_'s answer via the signaling protocol stream and sets it locally via `RTCPeerConnection.setRemoteDescription`.
+
+5. _B_ and _A_ send their local ICE candidates via the existing signaling protocol stream.
+   Both nodes continuously read from the stream, adding incoming remote candidates via `RTCPeerConnection.addIceCandidate()`.
+
+6. On successful establishment or failure of the direct connection, _B_ and _A_ close the signaling protocol stream.
+
+   Behavior for transferring data on a relayed connection, in the case where the direct connection failed, is out of scope for this specification and dependent on the application.
+
+7. Messages on `RTCDataChannel`s on the established `RTCPeerConnection` are framed using the message framing mechanism described in [multiplexing].
+
+## STUN
+
+A node needs to discover its public IP and port, which is forwarded to the remote node in order to connect to the local node.
+On non-browser libp2p nodes doing a hole punch with TCP or QUIC, the libp2p node discovers its public address via the [identify] protocol.
+One can not use the [identify] protocol on browser nodes to discover ones public IP and port given that the browser uses a new port for each connection.
+For example say that the local browser node establishes a WebRTC connection C1 via browser-to-server to a server node and runs the [identify] protocol.
+The returned observed public port P1 will most likely (depending on the NAT) be a different port than the port observed on another connection C2.
+The only browser supported mechanism to discover ones public IP and port for a given connection is the non-libp2p protocol STUN.
+This is why this specification depends on STUN, and thus the availability of one or more STUN servers for _A_ and _B_ to discovery their public addresses.
+
+There are various publicly available STUN servers.
+As an alternative one may operate dedicated STUN servers for a given libp2p network.
+Further specification of the usage of STUN is out of scope for this specifitcation.
+
+As an aside, note that _A_ and _B_ do not need to use the same STUN server in order to establish a direct WebRTC connection.
+
+## Signaling Protocol
+
+The protocol id is `/webrtc-direct`.
```

Comment:
```suggestion
The protocol id is `/webrtc-direct-signaling`.
```

Perhaps that would be clearer. I was a bit confused initially what protocol we were talking about here given that the components of a multiaddr are also called "protocols".

Reply:
Not a strong opinion:

I am in favor of appending `-signaling`.

I am in favor of removing `-direct`. In my eyes, the fact that a _direct_ connection is being established, is implied through `webrtc` already. In other words, I don't know of a case where a WebRTC connection would not be direct.

Proposal: `/webrtc-signaling`.

Happy to bikeshed further on this while other decisions are in-flight. That said, I don't think the decision on the protocol name should block the overall progress on this pull request.

Reply:
> Not a strong opinion:
> 
> I am in favor of appending `-signaling`.
> 
> I am in favor of removing `-direct`. In my eyes, the fact that a _direct_ connection is being established, is implied through `webrtc` already. In other words, I don't know of a case where a WebRTC connection would not be direct.

This could happen if a peer chooses to use a TURN server to relay UDP traffic, but it's not very common outside of companies using dedicated TURN servers for traffic on their network.

Reply:
Is there enough to have consensus on this proposed change?  @ckousik any objections?

Reply:
This should be okay for the stream protocol, but I feel we should advertise webrtc-signaling support in a multiaddr for ease of implementation until signed peer records have wider support. https://github.com/libp2p/specs/pull/497#discussion_r1107164657